dvc/cli.py

```diff
@@ -129,59 +129,48 @@ def __call__(self, parser, namespace, values, option_string=None):
         sys.exit(0)
 
 
-def get_parent_parser():
-    """Create instances of a parser containing common arguments shared among
-    all the commands.
+def add_common_args(parser, *, func=None):
+    """Add common arguments shared among all the commands."""
+    if func:
+        parser.set_defaults(func=func)
```

Comment:
This part is not really related to the common args, so passing `func` to `add_common_args` is confusing, as the name doesn't really reflect the action that well.

An alternative spin on that would be to just call `add_common_args(parser)` and not mess with `set_defaults`, but that will still require us calling that helper in every command. Which seems ok to me if there is no better way to do it.

Considering that all we want to do is make -h/-v appear last in the help section, maybe there are some argparse mechanisms we could leverage? E.g. maybe we could alter the way it formats the help in our DvcParser? It might turn out impossible or too hacky, but just want to consider it as it seems to have some potential to keep things localized and avoid us relying on the order in which we add things in the code.

Reply:
> [Explicit is better than implicit.](https://www.python.org/dev/peps/pep-0020/)

This is an explicit solution. It uses `agrparse` as it was designed. `agrparse` preserves the order of how you add your arguments and show them in that order, so it adds `-h` in init method and handles `parents` here as well:
https://github.com/python/cpython/blob/master/Lib/argparse.py#L1732-L1747
all later calls to `add_argument` will go down the list.

Now I do it in `agrparse` way and add these agrguments at the end, so it can be displayed as we want.

I don't understand why I should think about some magic way of just ordering some arguments differently. Yes, we can try to reorder [_ActionsContainer._actions](https://github.com/python/cpython/blob/fa1d83db62a545580d9a1a585b2c1fb55961a5c3/Lib/argparse.py#L1451), but it would be hacky and magical as you said, I'm not interested in implementing such a solution.

----

I agree on the name is confusing.

> An alternative spin on that would be to just call `add_common_args(parser)` and not mess with `set_defaults`

If you look at the commit history:
 [Move "parser.set_defaults(func=...)" to "add_common_args" as well](https://github.com/iterative/dvc/pull/4681/commits/ad86cbd64a3fa95b8b0099bfa8f59cfb99e2230a)
this move was in the last commit as I have seen repeatable two lines like
```py
    commit_parser.set_defaults(func=CmdCommit)
    add_common_args(commit_parser)
```

>  but that will still require us calling that helper in every command.

We already have assumptions for every command:
- we should call `add_parser` with `parents`
- we should call `parser.set_defaults(func=...)` as well

-----

I see a few options for this PR:
1. I revert my last commit with moving `parser.set_defaults(func=...)` into `add_common_args`
2. I rename `add_common_args` and keep `parser.set_defaults(func=...)` in it, maybe `add_common_parts`, or any better name
3. I just close this PR.

I'm fine with every option.

Reply:
Don't get me wrong, I'm all the way for explicit solutions :slightly_smiling_face:

I totally see your points :+1: , but just one more thing: you've probably considered this already, but just wanted to ask if you've considered using argparser's `formatting_class`?

Reply:
Btw, one more thing that I also got confused by: the ticket is about formatting in general and dvc.org seems to have a simple alphabetically sorted list of options, which we could do in dvc as well. Though the implementation for both approaches could be very similar. :slightly_smiling_face:

Reply:
@efiop please check my alternate solution here: #4695

Reply:
@naspeh Looks pretty good! :slightly_smiling_face: 

So have you considered `formatting_class`? It is an explicit potential solution that is provided by argparse specifically to customize help output.

Reply:
I have checked `formatter_class` option, but couldn't get quick result, we have subparsers-subparsers-subparser and you need to traverse all of these deep nestings to handle it properly. Haven't seen a near solution and come up with another one.

Reply:
@naspeh Got it, makes sense :ok_hand: 

It is important for this to be easy to maintain too, so a format_class-based solution seems promising. Seems like we could override `HelpFormatter.add_arguments` and either sort actions or just move -v/-h/-q to the end. I would probably go with sorting since it makes more sense. That way we won't need to override unrelated methods like we do in https://github.com/iterative/dvc/pull/4695 , would keep the original code and would format help where it is expected - in help formatter. What do you think?

Reply:
Closing both my PRs. Up to you.

Reply:
@naspeh Thanks for the PRs! :pray: